velox/common/compression/Compression.h

```diff
@@ -45,6 +48,208 @@ CompressionKind stringToCompressionKind(const std::string& kind);
 
 constexpr uint64_t DEFAULT_COMPRESSION_BLOCK_SIZE = 256 * 1024;
 
+static constexpr int32_t kUseDefaultCompressionLevel =
+    std::numeric_limits<int32_t>::min();
+
+class StreamingCompressor {
+ public:
+  virtual ~StreamingCompressor() = default;
+
+  struct CompressResult {
+    uint64_t bytesRead;
+    uint64_t bytesWritten;
+    bool outputTooSmall;
+  };
+
+  struct FlushResult {
+    uint64_t bytesWritten;
+    bool outputTooSmall;
+  };
+
+  struct EndResult {
+    uint64_t bytesWritten;
+    bool outputTooSmall;
+  };
+
+  /// Compress some input.
+  /// If CompressResult.outputTooSmall is true on return, compress() should be
+  /// called again with a larger output buffer, such as doubling its size.
+  virtual Expected<CompressResult> compress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// Flush part of the compressed output.
+  /// If FlushResult.outputTooSmall is true on return, flush() should be called
+  /// again with a larger output buffer, such as doubling its size.
+  virtual Expected<FlushResult> flush(
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// End compressing, doing whatever is necessary to end the stream, and
+  /// flushing the compressed output.
+  /// If EndResult.outputTooSmall is true on return, end() should be called
+  /// again with a larger output buffer, such as doubling its size.
+  /// Otherwise, the StreamingCompressor should not be used anymore.
+  virtual Expected<EndResult> end(uint8_t* output, uint64_t outputLength) = 0;
+};
+
+class StreamingDecompressor {
+ public:
+  virtual ~StreamingDecompressor() = default;
+
+  struct DecompressResult {
+    uint64_t bytesRead;
+    uint64_t bytesWritten;
+    bool outputTooSmall;
+  };
+
+  /// Decompress some input.
+  /// If DecompressResult.outputTooSmall is true on return, decompress() should
+  /// be called again with a larger output buffer, such as doubling its size.
+  virtual Expected<DecompressResult> decompress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  // Return whether the compressed stream is finished.
+  virtual bool isFinished() = 0;
+
+  // Reinitialize decompressor, making it ready for a new compressed stream.
+  virtual Status reset() = 0;
+};
+
+struct CodecOptions {
+  int32_t compressionLevel;
+
+  CodecOptions(int32_t compressionLevel = kUseDefaultCompressionLevel)
+      : compressionLevel(compressionLevel) {}
+
+  virtual ~CodecOptions() = default;
+};
+
+class Codec {
+ public:
+  virtual ~Codec() = default;
+
+  // Create a kind for the given compression algorithm with CodecOptions.
+  static Expected<std::unique_ptr<Codec>> create(
+      CompressionKind kind,
```

Comment:
could this be a template parameter?

Reply:
There are other static member functions take `CompressionKind` as the parameter. Do we need to change all of them?

Reply:
I think so. At least the ones that are not virtual, since you could remove the runtime branches. We can do that in another PR though to unblock this one.

Reply:
@pedroerp It requires more modifications to the UT. Can I address this in the next PR? Thanks!